### PR TITLE
Refactor: remove allocate_psi_init, initialize_psi

### LINKED ISSUE #4401: Function allocate_psi_init and initialize_psi should be remove from esolver_ks_pw.cpp
### Describe the Code Quality Issue

Esolver_KS_PW defined too much functions and they should be moved to other places.

### Additional Context

_No response_

### Task list for Issue attackers (only for developers)

- [ ] Identify the specific code file or section with the code quality issue.
- [ ] Investigate the issue and determine the root cause.
- [ ] Research best practices and potential solutions for the identified issue.
- [ ] Refactor the code to improve code quality, following the suggested solution.
- [ ] Ensure the refactored code adheres to the project's coding standards.
- [ ] Test the refactored code to ensure it functions as expected.
- [ ] Update any relevant documentation, if necessary.
- [ ] Submit a pull request with the refactored code and a description of the changes made.

### PR DESCRIPTION
**Pull request by Qianruipku**
### Linked Issue
Fix #4401

### What's changed?
- Remove allocate_psi_init, initialize_psi to InitController class.

### Next step
- I will delete wavefunc class and all things in wavefunc class should be controlled by psi_initailizer and InitController class.

### CODE REVIEW

File: source/module_hamilt_pw/hamilt_pwdft/initcontroller.cpp [hunk 1]
```diff
@@ -0,0 +1,256 @@
+#include "initcontroller.h"
```
Comment (mohanchen) [inline source/module_hamilt_pw/hamilt_pwdft/initcontroller.cpp]: how about change the file name to psi_init ?

File: source/module_hamilt_pw/hamilt_pwdft/initcontroller.h [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+                   wavefunc* p_wf_in,
+                   ModulePW::PW_Basis_K* pw_wfc_in);
+    ~InitController(){};
+
+    // prepare the wavefunction initialization
+    void prepare_init(Structure_Factor* p_sf, //< structure factor
+                      UnitCell* p_ucell,      //< unit cell
+                      const int& random_seed, //< random seed
+#ifdef __MPI
+                      pseudopot_cell_vnl* = nullptr, //< nonlocal pseudopotential
```
Comment (mohanchen) [inline source/module_hamilt_pw/hamilt_pwdft/initcontroller.h]: this line seems to be the same with or without MPI

File: source/module_esolver/esolver_ks_pw.h [hunk 3]
```diff
@@ -56,23 +56,14 @@ class ESolver_KS_PW : public ESolver_KS<T, Device>
...
     psi::Psi<std::complex<double>, base_device::DEVICE_CPU>* psi = nullptr;
 
-  private:
-    // psi_initializer<T, Device>* psi_init = nullptr;
-    // change to use smart pointer to manage the memory, and avoid memory leak
-    // while the std::make_unique() is not supported till C++14,
-    // so use the new and std::unique_ptr to manage the memory, but this makes new-delete not symmetric
-    std::unique_ptr<psi_initializer<T, Device>> psi_init;
+    // psi_initializer controller
+    psi::InitController<T, Device>* p_initcontroller = nullptr;
```
Comment (mohanchen) [inline source/module_esolver/esolver_ks_pw.h]: how about change p_initcontroller to p_init_psi or other names that are shorter?

File: source/module_hamilt_pw/hamilt_pwdft/initcontroller.cpp (hunk 1 above)
Comment (Qianruipku) [reply source/module_hamilt_pw/hamilt_pwdft/initcontroller.cpp]: In fact, we have class named psi_initializer and Initcontroller class has a member of "psi_initializer* psi_init". So it is not good to change the name to psi_init. However, the name "initcontroller" lacks the information of psi. It is not good, either.

Comment (Qianruipku) [reply source/module_hamilt_pw/hamilt_pwdft/initcontroller.cpp]: How about wfinit?

**Review by mohanchen — APPROVED** [review]

**Pull request merged**